baseplate/lib/propagator_redditb3.py

```diff
@@ -0,0 +1,129 @@
+import typing
+from re import compile as re_compile
+
+from opentelemetry import trace
+from opentelemetry.context import Context
+from opentelemetry.propagators.textmap import (
+    CarrierT,
+    Getter,
+    Setter,
+    TextMapPropagator,
+    default_getter,
+    default_setter,
+)
+from opentelemetry.trace import format_span_id, format_trace_id
+
+
+class RedditB3Format(TextMapPropagator):
+    """Propagator for the Reddit B3 HTTP header format."""
+
+    TRACE_ID_KEY = "X-Trace"
+    SPAN_ID_KEY = "X-Span"
+    SAMPLED_KEY = "X-Sampled"
+    FLAGS_KEY = "X-Flags"
+    _SAMPLE_PROPAGATE_VALUES = {"1", "True", "true", "d"}
+    # Although Reddit B3 uses 64bit TraceId's (16 char) we will accept 32 and truncate only when writing client headers.
+    _trace_id_regex = re_compile(r"[\da-fA-F]{16}|[\da-fA-F]{32}")
+    _span_id_regex = re_compile(r"[\da-fA-F]{16}")
+
+    def extract(
+        self,
+        carrier: CarrierT,
+        context: typing.Optional[Context] = None,
+        getter: Getter = default_getter,
+    ) -> Context:
+        if context is None:
+            context = Context()
+        trace_id = trace.INVALID_TRACE_ID
+        span_id = trace.INVALID_SPAN_ID
+        sampled = "0"
+        flags = None
+
+        trace_id = (
+            _extract_first_element(getter.get(carrier, self.TRACE_ID_KEY))
+            or trace_id
+        )
+        span_id = (
+            _extract_first_element(getter.get(carrier, self.SPAN_ID_KEY))
+            or span_id
+        )
+        sampled = (
+            _extract_first_element(getter.get(carrier, self.SAMPLED_KEY))
+            or sampled
+        )
+        flags = (
+            _extract_first_element(getter.get(carrier, self.FLAGS_KEY))
+            or flags
+        )
+
+        if (
+            trace_id == trace.INVALID_TRACE_ID
+            or span_id == trace.INVALID_SPAN_ID
+            or self._trace_id_regex.fullmatch(trace_id) is None
+            or self._span_id_regex.fullmatch(span_id) is None
+        ):
+            return context
```

Comment:
Can you add a comment that explains why / what's happening there for future reference?